### PR TITLE
Fix SPM build system

### PR DESCRIPTION
**Pull request by orenc17**
This changes the behaviour of our code generator.
code generation is split into 2 phases:
1. generate partition code ( for all manifests found in working tree)
2. generate common code ( only for manifests that are in the build proccess)

* psa_client_common.h was removed in the process, its replaced with partition headers describing SFIDs

### CODE REVIEW

**Comment by orenc17** [comment]
A passing jenkins run with the changes from ARMmbed/mbed-spm#85 can be found [here](https://jenkins-internal.mbed.com/view/mbed-spm/job/kfnta-mbed-os-new-autogen/3/)

**Comment by orenc17** [comment]
[latest jenkins run ](https://jenkins-internal.mbed.com/view/mbed-spm/job/kfnta-mbed-os-new-autogen/8/)

**Comment by orenc17** [comment]
[jenkins run after commits reordering](https://jenkins-internal.mbed.com/view/mbed-spm/job/kfnta-mbed-os-new-autogen/9/)

File: tools/spm/generate_partition_code.py [hunk 1]
```diff
@@ -465,93 +604,74 @@ def generate_source_files(
...
+            manifest.find_dependencies(manifests)
+        ))
 
+    return generated_folders
 
-def process_manifest_files(manifest_files, output_dir):
-    """
-    Process all the given manifest files and generate C code from them
 
+def generate_spm_data(manifest_files, output_dir, extra_filters=None):
```
Comment (mikisch81) [inline tools/spm/generate_partition_code.py]: I think this function and `generate_common_source_files()` can be in a single function which is called `generate_common_sources()`, and `process_manifest_files()` can be renamed to `generate_partitions_sources`.
This was my take on the naming comeete..

File: tools/spm/generate_partition_code.py [hunk 2]
```diff
@@ -316,6 +326,83 @@ def from_json(cls, manifest_file):
...
+        env = Environment(
+            loader=FileSystemLoader(templates_dirs),
+            lstrip_blocks=True,
+            trim_blocks=True
+        )
+        if extra_filters:
+            env.filters.update(extra_filters)
+
+        rendered_files = {}
+        output_dir = os.path.dirname(self.file)
```
Comment (mikisch81) [inline tools/spm/generate_partition_code.py]: We discussed that `output_dir` will be something like `psa_partition_autogen` or whatever, so it can be ignored properly by git.

File: tools/spm/generate_partition_code.py (hunk 1 above)
Comment (orenc17) [reply tools/spm/generate_partition_code.py]: converted `process_manifest_files()` -> `generate_partitions_sources()`

merging `generate_common_source_files()` and `generate_spm_data()` is problematic due to our current testing scheme

Comment (mikisch81) [reply tools/spm/generate_partition_code.py]: > merging `generate_common_source_files()` and `generate_spm_data()` is problematic due to our current testing scheme

Why?

File: tools/spm/partition_description_schema.json [hunk 3]
```diff
@@ -144,12 +144,16 @@
       }
     },
     "secure_func": {
-      "required": ["sfid", "non_secure_clients", "signal"],
+      "required": ["name", "identifier", "non_secure_clients", "signal"],
```
[the remark this answers is not in the document]
Comment (adrianlshaw) [reply tools/spm/partition_description_schema.json]: Hi @orenc17 @mikisch81 

After thinking for a bit, your proposal make sense. There were reasons why the SFID number wasn't specified in the original manifest (we were thinking of cases for dynamic allocation from outside the manifest). In PSA 0.6 we are now allowing manual allocation of the SFID, and it makes sense to do this in the manifest itself. I am currently writing the section about it. 

I suggest that the "identifier" field be of type "positive_integer_or_hex_string" (if it isn't already).

**Comment by orenc17** [comment]
[jenkins run with latest commits](https://jenkins-internal.mbed.com/view/mbed-spm/job/kfnta-mbed-os-new-autogen/12/)

File: tools/spm/generate_partition_code.py [hunk 4]
```diff
@@ -613,31 +581,28 @@ def generate_common_source_files(
...
-
-        full_path_common = path_join(output_folder, generated_filename)
-        rendered_files[full_path_common] = template.render(
-            partitions=manifests,
-            region_pair_list=region_pairs_list
-        )
+    for tf in template_files:
+        template = env.get_template(tf)
+        generated_filename = os.path.basename(templates[template_files[tf]])
+        rendered_files[generated_filename] = [templates[template_files[tf]], template.render(**render_args)]
```
[the remark this answers is not in the document]
Comment (orenc17) [reply tools/spm/generate_partition_code.py]: first elemnt is the full path of the generated file
second element is the render itself

this ugly change is due to the tests

File: tools/export/__init__.py [hunk 5]
```diff
@@ -75,7 +75,7 @@
 """
 
 def mcu_ide_list():
-    """Shows list of exportable ides
```
Comment (alzix) [inline tools/export/__init__.py]: please avoid unnecessary changes

File: tools/spm/generate_partition_code.py [hunk 6]
```diff
@@ -316,6 +316,86 @@ def from_json(cls, manifest_file):
...
+        :param generated_files: List of the generated files
+        :return: True/False
+        """
+
+        return is_up_to_date([self.file], template_files, generated_files)
+
+    def find_dependencies(self, manifests):
+        manifests = filter(lambda man: man != self, manifests)
+        extern_sfids_set = set(self.extern_sfids)
+        return [manifest.name for manifest in manifests if extern_sfids_set.intersection(set(manifest.sfids))]
```
Comment (alzix) [inline tools/spm/generate_partition_code.py]: i think we deserve an explanation here :)

File: tools/spm/generate_partition_code.py [hunk 7]
```diff
@@ -428,33 +508,73 @@ def get_latest_timestamp(*files):
...
 
 
-def generate_source_files(
+def is_up_to_date(manifest_files, template_files, generated_files):
+    """
+    Check if need to re-generate source files
+
+    We don't need to re-generate the files if:
+    1. All the expected generated files exist AND
+    2. None of the XMLs, templates, script files were modified after the
```
Comment (alzix) [inline tools/spm/generate_partition_code.py]: XML?
isn't it code duplication? we already have this function in the same file

File: TESTS/spm/client_tests/client_ipc_tests.cpp [hunk 8]
```diff
@@ -22,7 +22,7 @@
 #include "unity.h"
 #include "utest.h"
 #include "spm_client.h"
-#include "psa_client_common.h"
+#include "psa_client_tests_part1_partition_secure_functions.h"
```
Comment (alzix) [inline TESTS/spm/client_tests/client_ipc_tests.cpp]: do we really need such a long file names?
if i get it correctly `partition_secure_functions.h` is common suffix/
please consider to shorten it to `ifs.h` - secure functions are just an interfaces after all. and the name `secure function` term will be renamed in closest future.

File: tools/spm/partition_description_schema.json [hunk 9]
```diff
@@ -144,12 +144,16 @@
...
       "properties": {
-        "sfid": {
+        "name": {
           "description": "Alphanumeric C macro for referring to a Secure Function. (all capital)",
           "$ref": "#/definitions/c_macro"
         },
+        "identifier": {
+          "description": "Numeric value representing the secure function.",
+          "$ref": "#/definitions/hex_string"
```
Comment (alzix) [inline tools/spm/partition_description_schema.json]: please follow Adrian's suggestion and change the type to `positive_integer_or_hex_string`
